Client/Frontend/Settings/SettingsTableViewController.swift

```diff
@@ -194,6 +194,135 @@ class BoolSetting: Setting {
+
+    override func onConfigureCell(cell: UITableViewCell) {
+        super.onConfigureCell(cell)
+        if let id = accessibilityIdentifier {
+            textField.accessibilityIdentifier = id + "TextField"
+        }
+        textField.placeholder = placeholder
+        textField.delegate = self
+        textField.addTarget(self, action: #selector(textFieldDidChange), forControlEvents: .EditingChanged)
+
+        cell.accessibilityTraits = UIAccessibilityTraitLink
+        cell.contentView.addSubview(textField)
+
+        let container = UIView()
+        container.addSubview(textField)
+        cell.contentView.addSubview(container)
+
+        cell.contentView.snp_makeConstraints { make in
+            make.height.equalTo(44)
+            make.width.equalTo(cell.snp_width)
+        }
+
+        textField.snp_makeConstraints { make in
+            make.height.equalTo(cell.contentView)
+            make.width.equalTo(cell.contentView).offset(-2 * SettingsTableSectionHeaderFooterViewUX.titleHorizontalPadding)
+            make.leading.equalTo(cell.contentView).offset(SettingsTableSectionHeaderFooterViewUX.titleHorizontalPadding)
+        }
+        textField.text = prefs.stringForKey(prefKey) ?? defaultValue
+        textFieldDidChange(textField)
+    }
+
+    override func onClick(navigationController: UINavigationController?) {
+        textField.becomeFirstResponder()
+    }
+
+    private func isValid(value: String?) -> Bool {
+        return settingIsValid?(value) ?? true
+    }
+
+    @objc func textFieldDidChange(textField: UITextField) {
+        let color = isValid(textField.text) ? UIConstants.TableViewRowTextColor : UIConstants.DestructiveRed
+        textField.textColor = color
+    }
+
+    @objc func textFieldShouldReturn(textField: UITextField) -> Bool {
+        return isValid(textField.text)
+    }
+
+    @objc func textFieldDidEndEditing(textField: UITextField) {
+        let text = textField.text
+        if !isValid(text) {
+            return
+        }
+        // unsure if we want to actually set the value to the default here.
+        if let string = text ?? defaultValue {
+            prefs.setString(string, forKey: prefKey)
+            settingDidChange?(textField.text)
+        } else {
+            prefs.removeObjectForKey(prefKey)
+            settingDidChange?(defaultValue)
+        }
+    }
+}
+
+// A helper class for a setting backed by a UITextField.
+// This takes an optional isEnabled and mandatory onClick callback
+// isEnabled is called on each tableview.reloadData. If it returns
+// false then the 'button' appears disabled.
+class ButtonSetting: Setting {
+    let onButtonClick: (UINavigationController?) -> ()
+    let destructive: Bool
+    let isEnabled: () -> Bool
+
+    static let always = { return true }
+
+    init(title: NSAttributedString?, destructive: Bool = false, accessibilityIdentifier: String? = nil, isEnabled: () -> Bool = ButtonSetting.always, onClick: (UINavigationController?) -> ()) {
```

Comment:
In `StringSetting`, the `settingIsValid` parameter is defined as `(String? -> Bool)? = nil`. Why not do the same here for `isEnabled`? Then you won't need this `always` default argument.